conmon-rs/server/src/child_reaper.rs

```diff
@@ -0,0 +1,156 @@
+        let wait_lock = Arc::clone(&self.wait_lock);
+        thread::spawn(move || {
+            loop {
+                // To prevent an error: "No child processes (os error 10)" when spawning new runtime processes,
+                // we don't call waitpid until the runtime process has finished
+                // Inspired by
+                // https://github.com/kata-containers/kata-containers/blob/828a304883a4/src/agent/src/signal.rs#L27
+                let wait_status = {
+                    let wait_lock = Arc::clone(&wait_lock);
+                    let _lock = wait_lock.lock();
+                    waitpid(Pid::from_raw(-1), None)
+                };
+
+                let grandchildren = Arc::clone(&grandchildren);
+                match wait_status {
+                    Ok(status) => {
+                        if let WaitStatus::Exited(grandchild_pid, exit_status) = status {
+                            let grandchildren = Arc::clone(&grandchildren);
+                            thread::spawn(move || {
+                                if let Err(e) = Self::forget_grandchild(
+                                    grandchildren,
+                                    grandchild_pid,
+                                    exit_status,
+                                ) {
+                                    error!(
+                                        "Failed to reap grandchild for pid {}: {}",
+                                        grandchild_pid, e
+                                    );
+                                }
+                            });
+                        }
+                    }
+
+                    Err(err) => {
+                        // TODO FIXME this busy loops right now while there are no grandchildren.
+                        // There should be a broadcast mechanism so we only run this loop
+                        // while there are grandchildren
+                        if err != nix::errno::Errno::ECHILD {
+                            error!("caught error in reading for sigchld {}", err);
+                        }
+                    }
+                }
+            }
+        });
+        Ok(())
+    }
+
+    pub fn create_child<P, I, S>(
+        &self,
+        cmd: P,
+        args: I,
+    ) -> Result<std::process::ExitStatus, std::io::Error>
+    where
+        P: AsRef<std::ffi::OsStr>,
+        I: IntoIterator<Item = S>,
+        S: AsRef<std::ffi::OsStr>,
+    {
+        let mut cmd = std::process::Command::new(cmd);
+        cmd.args(args);
+
+        // To prevent an error: "No child processes (os error 10)",
+        // we prevent our reaper thread from calling waitpid until this child has completed.
+        // Inspired by
+        // https://github.com/kata-containers/kata-containers/blob/828a304883a4/src/agent/src/signal.rs#L27
+        let wait_lock = Arc::clone(&self.wait_lock);
+        let _lock = wait_lock.lock().unwrap();
+
+        cmd.spawn()?.wait()
+    }
```

Comment:
I think this can be changed to avoid the `unwrap`. Also, I'm not sure if we really need the `let wait_lock = Arc::clone(&self.wait_lock);`:

```diff
diff --git a/conmon-rs/server/src/child_reaper.rs b/conmon-rs/server/src/child_reaper.rs
index d44756c9..4264ea7f 100644
--- a/conmon-rs/server/src/child_reaper.rs
+++ b/conmon-rs/server/src/child_reaper.rs
@@ -1,6 +1,6 @@
 //! Child process reaping and management.
 
-use anyhow::{bail, Result};
+use anyhow::{bail, format_err, Context, Result};
 use getset::Getters;
 use log::{debug, error};
 use nix::sys::signal::{kill, Signal};
@@ -60,11 +60,7 @@ impl ChildReaper {
         Ok(())
     }
 
-    pub fn create_child<P, I, S>(
-        &self,
-        cmd: P,
-        args: I,
-    ) -> Result<std::process::ExitStatus, std::io::Error>
+    pub fn create_child<P, I, S>(&self, cmd: P, args: I) -> Result<std::process::ExitStatus>
     where
         P: AsRef<std::ffi::OsStr>,
         I: IntoIterator<Item = S>,
@@ -77,10 +73,12 @@ impl ChildReaper {
         // we prevent our reaper thread from calling waitpid until this child has completed.
         // Inspired by
         // https://github.com/kata-containers/kata-containers/blob/828a304883a4/src/agent/src/signal.rs#L27
-        let wait_lock = Arc::clone(&self.wait_lock);
-        let _lock = wait_lock.lock().unwrap();
+        let _ = &self
+            .wait_lock
+            .lock()
+            .map_err(|e| format_err!("take wait lock: {}", e))?;
 
-        cmd.spawn()?.wait()
+        Ok(cmd.spawn()?.wait().context("wait for child")?)
     }
 
     pub fn watch_grandchild(&self, id: String, pid: i32, exit_paths: Vec<PathBuf>) -> Result<()> {
```